### PR TITLE
feat: custom project root setting

### PR DESCRIPTION
**Pull request by kasiaMarek**

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala [hunk 1]
```diff
@@ -121,21 +121,40 @@ final class BuildTools(
...
-        .collectFirst {
-          case file
-              if file.isDirectory &&
-                !file.getName.startsWith(".") &&
-                isProjectRoot(AbsolutePath(file.toPath())) =>
-            AbsolutePath(file.toPath())
-        }
+  ): Option[AbsolutePath] = {
+    customProjectRoot match {
+      case Some(projectRoot) => Some(projectRoot).filter(isProjectRoot)
```
Comment (jkciesluk) [inline metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala:142]:
```suggestion
      case Some(projectRoot) if isProjectRoot(projectRoot) => Some(projectRoot)
```

File: metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala [hunk 2]
```diff
@@ -121,21 +121,40 @@ final class BuildTools(
   )
   def isBazel: Boolean = bazelProject.isDefined
 
+  private def customProjectRoot =
+    userConfig().customProjectRoot
+      .map(_.trim() match {
+        case "." => workspace
+        case relativePath => workspace.resolve(relativePath)
```
Comment (jkciesluk) [inline metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala:128]: Just to make sure, if customProjectRoot is empty (`Some("")`) it resolves to workspace here and works as if it was `None`?

File: metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala (hunk 1 above)
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala:142]: It's on purpose like this. So if you choose a custom root it should look only in that root. We had a problem discussed on discord, where a person had a custom bsp and an unused `sbt`  somewhere lower file hierarchy.

File: metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala (hunk 2 above)
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala:128]: Empty is filtered out while `fromJson` on the whole `UserConfig`.

File: metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala [hunk 3]
```diff
@@ -313,6 +314,14 @@ object UserConfiguration {
           |launcher, not available in PATH.
           |""".stripMargin,
       ),
+      UserConfigurationOption(
+        "custom-project-root",
+        """empty string `""`.""",
+        """"backend/scalaProject/"""",
+        "Scala CLI launcher",
```
Comment (tgodzik) [inline metals/src/main/scala/scala/meta/internal/metals/UserConfiguration.scala:321]:
```suggestion
        "Custom project root",
```

**Review by tgodzik — APPROVED** [review]
Looks good! Just two minor comments

File: metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (kasiaMarek) [reply metals/src/main/scala/scala/meta/internal/builds/BuildTools.scala:128]: I wasn't sure about Windows.

**Pull request merged**